Content.Shared/Item/ItemToggle/SharedItemToggleSystem.cs

```diff
@@ -0,0 +1,313 @@
+using Content.Shared.Interaction.Events;
+using Content.Shared.Toggleable;
+using Content.Shared.Weapons.Melee;
+using Content.Shared.Temperature;
+using Content.Shared.Wieldable;
+using Content.Shared.Wieldable.Components;
+using Robust.Shared.Audio;
+using Robust.Shared.Audio.Systems;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Network;
+
+namespace Content.Shared.Item.ItemToggle;
+/// <summary>
+/// Handles generic item toggles, like a welder turning on and off, or an e-sword.
+/// </summary>
+/// <remarks>
+/// If you need extended functionality (e.g. requiring power) then add a new component and use events.
+/// </remarks>
+public abstract class SharedItemToggleSystem : EntitySystem
+{
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly SharedAppearanceSystem _appearance = default!;
+    [Dependency] private readonly SharedItemSystem _item = default!;
+    [Dependency] private readonly SharedPointLightSystem _light = default!;
+    [Dependency] private readonly INetManager _netManager = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<ItemToggleComponent, UseInHandEvent>(OnUseInHand);
+        SubscribeLocalEvent<ItemToggleHotComponent, IsHotEvent>(OnIsHotEvent);
+        SubscribeLocalEvent<ItemToggleComponent, ItemUnwieldedEvent>(TurnOffonUnwielded);
+        SubscribeLocalEvent<ItemToggleComponent, ItemWieldedEvent>(TurnOnonWielded);
+        SubscribeLocalEvent<ItemToggleComponent, ItemToggleForceToggleEvent>(ForceToggle);
+    }
+    private void OnUseInHand(EntityUid uid, ItemToggleComponent itemToggle, UseInHandEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        args.Handled = true;
+
+        if (TryComp<WieldableComponent>(uid, out var wieldableComp))
+            return;
+
+        Toggle(uid, args.User, itemToggle);
+    }
+
+    public void Toggle(EntityUid uid, EntityUid? user = null, ItemToggleComponent? itemToggle = null)
+    {
+        if (!Resolve(uid, ref itemToggle))
+            return;
+
+        if (itemToggle.Activated)
+        {
+            TryDeactivate(uid, user, itemToggle: itemToggle);
+        }
+        else
+        {
+            TryActivate(uid, user, itemToggle: itemToggle);
+        }
+    }
+
+    public void ForceToggle(EntityUid uid, ItemToggleComponent itemToggle, ref ItemToggleForceToggleEvent args)
+    {
+        Toggle(uid, args.User, itemToggle);
+    }
+
+    public bool TryActivate(EntityUid uid, EntityUid? user = null, ItemToggleComponent? itemToggle = null)
+    {
+        if (!Resolve(uid, ref itemToggle))
+            return false;
+
+        if (itemToggle.Activated)
+            return true;
+
+        // The client cannot predict if the attempt to turn on fails or not since the battery and fuel systems are server side (for now). Potential future TODO
+        if (_netManager.IsServer)
+        {
```

Comment:
Should just add a bool arg for predicted that defaults to true instead as this is gonna feel pretty bad and newbies will just ignore prediction issues with their own code as a result.

Reply:
I've added that and changed up stuff as a result, but I've still got a bit of an issue here. The thing is that when you try to activate, say, a stun baton with 0 charge, because the stun baton system is server side, the server knows that the attempt failed, but the client does not. 
So how would I make it so the stun baton makes the failure to activate sound but not the spark sound?

Reply:
I guess I added a bool arg for predicted, and also added a predictable field in the itemToggle component (which defaults to true). That way if the item's toggle is unpredictable, the client won't run the code for activation stuff, and it gets the changes from the autonetworked fields.